src/Data/Graph/ControlFlowVertex.hs

```diff
@@ -99,63 +100,74 @@ instance ToJSON ControlFlowVertex where
 -- Typeclasses to create 'ControlFlowVertex's from 'Term's. Also extracts
 -- 'Name's for terms with symbolic names like Identifiers and Declarations.
 
-class VertexDeclaration syntax where
-  toVertex :: (Declarations1 syntax, Foldable syntax)
-           => Loc
-           -> ModuleInfo
-           -> syntax (Term syntax Loc)
-           -> Maybe (ControlFlowVertex, Name)
+class VertexDeclaration term where
+  toVertex
+    :: ModuleInfo
+    -> term Loc
+    -> Maybe (ControlFlowVertex, Name)
 
-instance (VertexDeclaration' syntax syntax) => VertexDeclaration syntax where
-  toVertex = toVertex'
+instance (VertexDeclaration1 f, Declarations1 f) => VertexDeclaration (Term f) where
+  toVertex info (Term (In a f)) = liftToVertex toVertex a info f
 
-class VertexDeclaration' whole syntax where
-  toVertex' :: (Declarations1 whole, Foldable whole)
-            => Loc
-            -> ModuleInfo
-            -> syntax (Term whole Loc)
-            -> Maybe (ControlFlowVertex, Name)
+instance (VertexDeclaration1 f, Declarations1 f) => VertexDeclaration (Quieterm f) where
+  toVertex info (Quieterm (In a f)) = liftToVertex toVertex a info f
 
-instance (VertexDeclarationStrategy syntax ~ strategy, VertexDeclarationWithStrategy strategy whole syntax) => VertexDeclaration' whole syntax where
-  toVertex' = toVertexWithStrategy (Proxy :: Proxy strategy)
+toVertex1 :: (VertexDeclaration1 f, VertexDeclaration t, Declarations (t Loc)) => Loc -> ModuleInfo -> f (t Loc) -> Maybe (ControlFlowVertex, Name)
+toVertex1 = liftToVertex toVertex
+
+class VertexDeclaration1 syntax where
+  liftToVertex :: Declarations (term Loc)
+               => (ModuleInfo -> term Loc -> Maybe (ControlFlowVertex, Name))
+               -> Loc
+               -> ModuleInfo
+               -> syntax (term Loc)
+               -> Maybe (ControlFlowVertex, Name)
+
+instance (VertexDeclarationStrategy1 syntax ~ strategy, VertexDeclarationWithStrategy1 strategy syntax) => VertexDeclaration1 syntax where
+  liftToVertex = liftToVertexWithStrategy (Proxy :: Proxy strategy)
+
+-- | This appears to be required to convince 'Semantic.Graph.runCallGraph' not to try to specialize the instance too eagerly.
+instance {-# OVERLAPPING #-} VertexDeclaration1 V1 where
+  liftToVertex _ _ _ v = case v of {}
```

Comment:
I don’t understand why it was so keen on specializing to the one instance, but this fixed it.

Reply:
Mysterious. Can we ask someone familiar with GHC internals?

Reply:
@kcsongor might know…

Reply:
What was the issue exactly? I had a quick look at `runCallGraph` but it wasn't immediately obvious where the `VertexDeclaration1` instance is picked up.

Reply:
@kcsongor: before defining this instance, some places where I had given a `VertexDeclaration1 t` constraint errored saying that it couldn’t satisfy the constraints from the context for the other instance (at that time the only instance)—i.e. it was trying to solve the sole instance’s context directly rather than using the given constraint, which surprised me.

This struck me as reminiscent of your _[Opaque constraint synonyms](https://kcsongor.github.io/opaque-constraint-synonyms/)_, so I thought you might know why it was solving _through_ the given constraint instead of just stopping with it 😊

Reply:
You might’ve already answered it in that blog post, actually:

> This happens because the compiler sees that there’s only one matching instance, so it’s safe to pick that one, and it will do so. This point is the important one: that there’s only one instance. So, if we could somehow trick GHC into thinking that there are other options, then maybe it wouldn’t be so eager to expand our constraints.

I was surprised it was applying this at a call site with a given constraint, but I guess if that function weren’t exported (I forget now if it was or not) then there’d be no way for any other instance to ever arrive there.

Reply:
I see! Yes, this happens with both givens and wanteds (as they get simplified in the same way), then the constraints of the given get turned into wanteds...